File: Data Structures & Algorithms/combination-target-sum/submission-1.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        all_combs = []
        # candidates.sort()
        def traverse(nums: List[int], start: int, curr_sum: int):
            if curr_sum == target:
                all_combs.append(nums.copy())
                return
            if curr_sum > target:
                return
            for i in range(start, len(candidates)):
                candidate = candidates[i]
                nums.append(candidate)
                traverse(nums, i, curr_sum + candidate)
                nums.pop()
        traverse([], 0, 0)
        return all_combs

File: Data Structures & Algorithms/three-integer-sum/submission-1.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[][]}
     */
    threeSum(nums) {
        const result = []
        nums.sort()
        for (let i=0;i<nums.length-2;i++) {
            if (i > 0 && nums[i-1] === nums[i]) {
                continue;
            }
            for (let j=i+1;j<nums.length-1;j++) {
                if (j > i+1 && nums[j-1] === nums[j]) {
                    continue;
                }
                for (let k=j+1;k<nums.length;k++) {
                    if (k > j+1 && nums[k-1] === nums[k]) {
                        continue;
                    }
                    if (nums[i] + nums[j] + nums[k] === 0) {
                        result.push([nums[i],nums[j],nums[k]])
                    }
                }
            }
        }
        return result;
    }
}
